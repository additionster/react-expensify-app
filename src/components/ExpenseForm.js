import React from 'react';
import moment from 'moment';
import DatePicker from 'react-datepicker';


class ExpenseForm extends React.Component {
    state = {
        description: this.props.expense? this.props.expense.description: '',
        amount: this.props.expense? (this.props.expense.amount/100).toString():'',
        note: this.props.expense? this.props.expense.note:'',
        createdAt: this.props.expense? moment(this.props.expense.createdAt):moment(),
        error: ''
    };
    onDescriptionChange = (e) => {
        const desc = e.target.value;
        this.setState(() => ({description:desc}));
    };
    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    };
    onCreatedAt = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({ createdAt }));
        }   
    };
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({note}));
    };
    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ error: 'Please provide description and amount.'}));
        }
        else {
            this.setState(() => ({ error: '' }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    };
    render() {
        return (
            <div>
                {this.state.error && <p style={{color:"red"}}>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input 
                        type="text"
                        placeholder='Description'
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input 
                        type="text"
                        placeholder='Amount'
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <DatePicker 
                        showIcon
                        showMonthDropdown
                        showYearDropdown
                        dropdownMode='select'
                        selected={this.state.createdAt}
                        minDate={moment().subtract(10,'days').calendar()}
                        dateFormat="dd/MM/yyyy"
                        placeholderText='dd/MM/yyyy'
                        onChange={this.onCreatedAt}
                    />
                    <textarea 
                        value={this.state.note}
                        placeholder='Add a note for your expense  (optional)'
                        onChange={this.onNoteChange}
                    >
                    </textarea>
                    <button>{this.state.description? "Edit":"Add"} Expense</button>
                </form>
            </div>
        );
    }
}

export default ExpenseForm;