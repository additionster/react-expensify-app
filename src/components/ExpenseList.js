import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses";

export const ExpenseList = (props) => (
    <div>
        <h1>Expense List</h1>
        {
            props.expenses && props.expenses.length !== 0? (
                props.expenses.map((expense) => {
                    return <ExpenseListItem key={expense.id} {...expense}/>
                })
            ) : (
                <p>No expenses</p>
            )
        }
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};
//Connect store from redux using HOC pattern
const ConnectedExpenseList = connect(mapStateToProps)(ExpenseList);

export default ConnectedExpenseList;