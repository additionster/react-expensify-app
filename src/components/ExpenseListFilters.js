import React from "react";
import { connect } from "react-redux";
import { setEndDate, setStartDate, setTextFilter, sortByAmount, sortByDate } from "../actions/filters";
import DatePicker from "react-datepicker";

export class ExpenseListFilters extends React.Component {

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    }

    onSortChange = (e) => {
        switch(e.target.value)
            {
                case 'date': this.props.sortByDate();break;
                case 'amount': this.props.sortByAmount();break;
            }
    }

    onDateChange = (dates) => {
        const [startDate, endDate] = dates;
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }

    render() {
        return (
            <div>
                <input type="text" defaultValue={this.props.filters.text} onChange={this.onTextChange} />
                <select defaultValue={this.props.filters.sortBy} onChange={this.onSortChange}>
                    <option value='date'>Date</option>
                    <option value='amount'>Amount</option>
                </select>
                <DatePicker 
                    dateFormat='dd/MM/yyyy'
                    selectsRange={true}
                    startDate={this.props.filters.startDate?.toDate()}
                    endDate={this.props.filters.endDate?.toDate()}
                    onChange={this.onDateChange}
                    placeholderText="startDate - endDate"
                    isClearable={true}
                />
            </div>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    };
};

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (value) => dispatch(setTextFilter(value)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))

});
const ConnectedExpenseListFilters = connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);

export default ConnectedExpenseListFilters;