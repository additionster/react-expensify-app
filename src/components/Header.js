import React from "react";
import { NavLink } from "react-router";
import { connect } from "react-redux";
import { startLogout } from "../actions/auth";
export const Header = ({ startLogout, isAuthenticated }) => (
    <header>
        <h1>Expensify</h1>
        <NavLink to="/dashboard" className={({isActive}) => isActive? "is-active":"in-active"}>Dashboard</NavLink> | 
        <NavLink to="/create" className={({isActive}) => isActive? "is-active":"in-active"}>Create Expense</NavLink> | 
        <NavLink to="/help" className={({isActive}) => isActive? "is-active":"in-active"}>Help</NavLink>
        {
            isAuthenticated &&
            <button onClick={startLogout}>Logout</button>
        }
    </header>
);

const mapStateToProps = (state) => ({
    isAuthenticated: !!state.auth.uid
});

const mapDispatchToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout())
});
export default connect(mapStateToProps, mapDispatchToProps)(Header);