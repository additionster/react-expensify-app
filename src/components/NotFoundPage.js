import React from "react";
import {Link} from 'react-router';

const NotFoundPage = () => (
    <div>
        Page not Found - <Link to="/" className='is-active'>Go Home</Link>
    </div>
);
export default NotFoundPage;