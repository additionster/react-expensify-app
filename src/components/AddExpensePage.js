import React from "react";
import { connect } from "react-redux";
import ExpenseForm from './ExpenseForm';
import { addExpense } from "../actions/expenses";
import { Navigate } from "react-router";


export class AddExpensePage extends React.Component {
    state = {
        submitted: false
    }
    onSubmit = (expense) => {
        //Cannot use hook in arrow function. so replace with <Navigate to/>
        //const navigate = useNavigate();
        this.props.addExpense(expense);
        //navigate('/');
        this.setState({submitted: true});
    }

    render() {
        return (
            <div>
                {this.state.submitted && <Navigate to="/" />}
                <h1>Add Expense</h1>
                <ExpenseForm 
                    onSubmit={this.onSubmit}
                />
            </div>
        );
    };
}

const mapDispatchToProps = (dispatch) => ({
    addExpense: (expense) => dispatch(addExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);