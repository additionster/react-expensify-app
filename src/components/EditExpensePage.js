import React, { Component } from "react";
import { connect } from "react-redux";
import ExpenseForm from './ExpenseForm';
import { Navigate, useParams } from "react-router";
import { editExpense, removeExpense } from "../actions/expenses";

//wrap hook (useParams) inside function based component
function withParams(Component) {
    return props => <Component {...props} params={useParams()} />;
}

export class EditExpensePage extends React.Component {
    state = {
        submitted: false
    }

    onSubmit = (expense) => {
        this.props.editExpense(this.props.params.id, expense);
        this.setState({submitted: true});
    }

    onRemove = () => {
        this.props.removeExpense(this.props.params.id);
        this.setState({submitted: true});
    }

    render() {
        return (
            <div>
                {this.state.submitted && <Navigate to="/" />}
                <h1>Edit Expense</h1>
                <ExpenseForm 
                    expense={this.props.expenses.find((expense) => expense.id === this.props.params.id)}
                    onSubmit={this.onSubmit}
                />
                <button onClick={this.onRemove}>
                    Remove
                </button>
            </div>
        );
    };
}

const mapStateToProps = (state) => {
    return {
        expenses: state.expenses
    };
};

const mapDispatchToProps = (dispatch) => ({
    editExpense: (id, expense) => dispatch(editExpense(id, expense)),
    removeExpense: (id) => dispatch(removeExpense({id}))
})

const connectedEditExpense = connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);

export default withParams(connectedEditExpense);