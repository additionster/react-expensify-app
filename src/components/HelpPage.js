import React from "react";
import Header from "./Header";
const HelpPage = () => (
    <div>
        <Header />
        This is from my help expense component
    </div>
);
export default HelpPage;