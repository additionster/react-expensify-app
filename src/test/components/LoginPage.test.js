import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import '@testing-library/jest-dom';
import { LoginPage } from "../../components/LoginPage";
import { BrowserRouter } from "react-router";

let startLoginButtonSpy, renderer;

beforeEach(() => {
    startLoginButtonSpy = jest.fn();
    renderer = render(
        <BrowserRouter>
            <LoginPage startLogin={startLoginButtonSpy}/>
        </BrowserRouter>
    );
});

test('testRenderLoginPage', () => {
    expect(renderer.baseElement).toMatchSnapshot();
});

test('shouldCallStartLoginButton', () => {
    const loginButton = screen.getAllByRole('button').at(0);
    fireEvent.click(loginButton);
    expect(startLoginButtonSpy).toHaveBeenCalled();
});