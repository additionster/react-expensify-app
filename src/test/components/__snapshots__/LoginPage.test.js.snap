// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`testRenderLoginPage 1`] = `
<body>
  <div>
    <div>
      <button>
        Login
      </button>
    </div>
  </div>
</body>
`;
